spec/lib/upload_creator_spec.rb

```diff
@@ -0,0 +1,69 @@
+require 'rails_helper'
+
+RSpec.describe UploadCreator do
+  let(:user) { Fabricate(:user) }
+
+  describe '#create_for' do
+    describe 'when upload is not an image' do
+      before do
+        SiteSetting.authorized_extensions = 'txt'
+      end
+
+      let(:filename) { "utf-8.txt" }
+      let(:file) { file_from_fixtures(filename, "encodings") }
+
+      it 'should store the upload with the right extension' do
+        expect do
+          UploadCreator.new(file, filename).create_for(user.id)
+        end.to change { Upload.count }.by(1)
+
+        upload = Upload.last
+
+        expect(upload.extension).to eq('txt')
+        expect(File.extname(upload.url)).to eq('.txt')
+        expect(upload.original_filename).to eq('utf-8.txt')
+      end
+    end
+
+    describe 'when image has the wrong extension' do
+      let(:filename) { "png_as.jpg"}
+      let(:file) { file_from_fixtures(filename) }
+
+      it 'should store the upload with the right extension' do
+        expect do
+          UploadCreator.new(file, filename).create_for(user.id)
+        end.to change { Upload.count }.by(1)
+
+        upload = Upload.last
+
+        expect(upload.extension).to eq('png')
+        expect(File.extname(upload.url)).to eq('.png')
+        expect(upload.original_filename).to eq('png_as.jpg')
```

Comment:
@SamSaffron I think we should keep the original filename as it is for debugging purposes. I found out about this issue by noticing that the original filename and extension was different.